sysl2/sysl/seqs/writer.go

```diff
@@ -0,0 +1,172 @@
+package seqs
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+)
+
+type SequenceDiagramWriter struct {
+	ind            int
+	atBeginOfLine  bool
+	autogenWarning bool
+	active         map[string]int
+	properties     []string
+	head           bytes.Buffer
+	body           bytes.Buffer
+}
+
+func MakeSequenceDiagramWriter(autogenWarning bool, properties ...string) *SequenceDiagramWriter {
+	p := make([]string, 0, len(properties))
+	p = append(p, properties...)
+	return &SequenceDiagramWriter{
+		atBeginOfLine:  true,
+		autogenWarning: autogenWarning,
+		properties:     p,
+		active:         make(map[string]int),
+	}
+}
+
+func (s *SequenceDiagramWriter) WriteTo(w io.Writer) (n int64, err error) {
+	i, err := fmt.Fprint(w, s.String())
+
+	return int64(i), err
+}
+
+func (s *SequenceDiagramWriter) Write(p []byte) (n int, err error) {
+	newline := []byte("\n")
+	newlines := bytes.Count(p, newline)
+	if newlines == 0 {
+		if s.atBeginOfLine {
+			s.writeIndent()
+		}
+		n, err = s.body.Write(p)
+		s.atBeginOfLine = false
+		return n, err
+	}
+
+	frags := bytes.SplitN(p, newline, newlines+1)
+
+	for i, frag := range frags {
+		if s.atBeginOfLine && len(frag) != 0 {
+			s.writeIndent()
+		}
+		nn, err := s.body.Write(frag)
+		s.atBeginOfLine = false
+		n += nn
+		if err != nil {
+			return n, err
+		}
+		if i+1 < len(frags) {
+			if err := s.WriteByte('\n'); err != nil {
+				return n, err
+			}
+			n++
+		}
+	}
+	s.atBeginOfLine = len(frags[len(frags)-1]) == 0
+	return n, nil
+}
+
+func (s *SequenceDiagramWriter) WriteString(v string) (n int, err error) {
```

Comment:
Does this need to handle `\n` in `v`?

Reply:
The actual write logic is in `Write([]byte) (int, error)`. This method just a wrapper of `Write([]byte) (int, error)` to handle the type conversion from `string` to `[]byte`